File: raycasting.py
from settings import *

class RayCasting:

    @staticmethod
    def cast_ray(pos, map_pos, angle, max_length, map_dic):
        sin_angle = math.sin(angle)
        cos_angle = math.cos(angle)

        #horizonal raycast
        hor_grid_coll_value = -1

        #determine which vertical direction the ray is going
        hor_y, hor_dy = (map_pos[1] + 1, 1) if sin_angle > 0 else (map_pos[1] - 1e-6, -1)

        #find inital depth and first y intercept x coordinate
        hor_depth = (hor_y - pos[1]) / sin_angle
        hor_x = pos[0] + hor_depth * cos_angle

        #find the depth delta of every y intecept jump and the delta x of every jump
        hor_depth_delta = hor_dy / sin_angle
        hor_dx = hor_depth_delta * cos_angle

        while (hor_depth < max_length):
            hor_map_pos = int(hor_x), int(hor_y)
            if (hor_map_pos in map_dic):
               hor_grid_coll_value = map_dic[hor_map_pos]
               break
            hor_x += hor_dx
            hor_y += hor_dy
            hor_depth += hor_depth_delta

            if (hor_depth > max_length):
               hor_depth = max_length

        #vertical raycast
        vert_grid_coll_value = -1

        #determine which vertical direction the ray is going
        vert_x, vert_dx = (map_pos[0] + 1, 1) if cos_angle > 0 else (map_pos[0] - 1e-6, -1)

        #find inital depth and first x intercept y coordinate
        vert_depth = (vert_x - pos[0]) / cos_angle
        vert_y = pos[1] + vert_depth * sin_angle

        #find the depth delta of every x intecept jump and the delta y of every jump
        vert_depth_delta = vert_dx / cos_angle
        vert_dy = vert_depth_delta * sin_angle

        while (vert_depth < max_length):
            vert_map_pos = int(vert_x), int(vert_y)
            if (vert_map_pos in map_dic):
                vert_grid_coll_value = map_dic[vert_map_pos]
                break
            vert_x += vert_dx
            vert_y += vert_dy
            vert_depth += vert_depth_delta

            if (vert_depth > max_length):
                vert_depth = max_length

        if (hor_depth == max_length and vert_depth == max_length):
            return (False, -1, None, -1)
        
        if (hor_depth <= vert_depth):
            return (True, hor_depth, (hor_x, hor_y), hor_grid_coll_value)
        else:
            return (True, vert_depth, (vert_x, vert_y), vert_grid_coll_value)

File: C_Files/raycasting.c
#include "raycasting.h"

Ray cast_ray(const Point *pos, const IntPoint *map_pos, const double angle, const double max_length) {
    double sin_angle = sin(angle);
    double cos_angle = cos(angle);

    int hor_grid_coll_value = -1;
    IntPoint hor_map_pos;

    double hor_y = (sin_angle > 0) ? (map_pos->y) + 1 : (map_pos->y - 1e-6);
    int hor_dy = (sin_angle > 0) ? 1 : -1;

    double hor_depth = (hor_y - pos->y) / sin_angle;
    double hor_x = pos->x + (hor_depth * cos_angle);

    double hor_depth_delta = hor_dy / sin_angle;
    double hor_dx = hor_depth_delta * cos_angle;

    while (hor_depth < max_length) {
        hor_map_pos.x = hor_x;
        hor_map_pos.y = hor_y;
        if (map1[hor_map_pos.y][hor_map_pos.x] != 0) {
            hor_grid_coll_value = map1[hor_map_pos.y][hor_map_pos.x];
            break;         
        }
        hor_x += hor_dx;
        hor_y += hor_dy;
        hor_depth += hor_depth_delta;

        if (hor_depth > max_length) { hor_depth = max_length; }
    }

    int vert_grid_coll_value = -1;
    IntPoint vert_map_pos;

    double vert_x = (cos_angle > 0) ? (map_pos->x + 1) : (map_pos->x - 1e-6);
    int vert_dx = (cos_angle > 0) ? 1 : -1;

    double vert_depth = (vert_x - pos->x) / cos_angle;
    double vert_y = pos->y + (vert_depth * sin_angle);

    double vert_depth_delta = vert_dx / cos_angle;
    double vert_dy = vert_depth_delta * sin_angle;

    while (vert_depth < max_length) {
        vert_map_pos.x = vert_x;
        vert_map_pos.y = vert_y;
        if (map1[vert_map_pos.y][vert_map_pos.x] != 0) {
            vert_grid_coll_value = map1[vert_map_pos.y][vert_map_pos.x];
            break;
        }
        vert_x += vert_dx;
        vert_y += vert_dy;
        vert_depth += vert_depth_delta;

        if (vert_depth > max_length) { vert_depth = max_length; }
    }

    Ray ray;
    if (hor_depth >= max_length && vert_depth >= max_length) {
        ray.hit = 0;
        ray.depth = -1;
        ray.hit_loc.x = -1;
        ray.hit_loc.y = -1;
        ray.grid_val = -1;
        return ray;
    }

    ray.hit = 1;
    if (hor_depth <= vert_depth) {
        ray.depth = hor_depth;
        ray.hit_loc = hor_map_pos;
        ray.grid_val = hor_grid_coll_value;
    } else {
        ray.depth = vert_depth;
        ray.hit_loc = vert_map_pos;
        ray.grid_val = vert_grid_coll_value;
    }
    return ray;
}